oak/server/rust/oak_runtime/src/runtime/mod.rs

```diff
@@ -679,6 +691,41 @@ impl Runtime {
         node_info.label.clone()
     }
 
+    /// Returns a version of the [`Label`] of the provided Node, taking into account all the
+    /// downgrade abilities that the node possesses, based on its set of
+    /// [`NodeInfo::can_downgrade_tags`].
+    fn get_node_downgraded_label(&self, node_id: NodeId) -> Label {
+        // Original (static) Node label.
+        let node_label = self.get_node_label(node_id);
+        // Retrieve the set of tags that the node may downgrade.
+        let node_can_downgrade_tags = self.get_node_can_downgrade_tags(node_id);
+        Label {
+            // Remove (declassify) secrecy tags that the Node can downgrade.
+            secrecy_tags: node_label
+                .secrecy_tags
+                .iter()
+                .filter(|t| !node_can_downgrade_tags.contains(t))
+                .cloned()
+                .collect(),
+            // Add (endorse) integrity tags that the Node can downgrade.
+            integrity_tags: node_label
+                .integrity_tags
+                .iter()
+                .chain(node_can_downgrade_tags.iter())
+                .cloned()
+                .collect(),
+        }
+    }
```

Comment:
@aferr do you think this logic (and associated comments) makes sense? The idea is that each node has a static (immutable) label, and a set of capabilities to downgrade specific tags (for now assume they are "dual" capabilities, as Flume calls them, i.e. they can be both endorsed and declassified). Here I am computing the "most public" label that the node can declassify to, and use that to perform the actual "flows to" comparison when reading from and writing to channels. Is this correct?

Reply:
Overall this does seem correct and like a fine way to do the downgrade check. The comments are also consistent with what the code does. Running through roughly what you said, if you wanted to have a top-level API like
downgrading_channel_send(self, node_id, channel_handle, ..., down_label)

roughly this would work like:
get_node_downgraded_label(self, nod_id) flowsTo down_label. /* to check that this downgrade is allowed */
down_label flows_to get_channel_label(channel_handle) /* to check that the downgrade is sufficient to do the send */

Another way to use this that I think is a bit less safe:
You could also just relax the check that gets done on the normal channel_send by checking that get_node_downgraded_label(self) flows_to get_ch_label(handle)
but I think this is less safe because then programmers will do downgrades when they don't realize they are doing it. I think it's better to make downgrades explicit because they are exceptions to isolation.

You could make this more general (And I recommend that you do!):
The idea is that if t is in node_can_downgrade_tags, that means it can either declassify away t from confidentiality or add t to integrity? This seems fine, but you might get better generality if you have separate capabilities for declassifying/endorsing (so there is node_can_declassify_tags and node_can_endorse_tags). It seems useful to be able to control declassifying/endorsing separately and basically free to allow this. The Flume paper actually does have separate capabilities for "adding labels" and "removing labels", which is a bit confusing because for confidentiality adding is upgrading and removing is downgrading. So in Flume, t- is the capability for declassifying and for reducing integrity (which is not endorsing!) and then t+ is endorsing/conf_upgrading (whacky!). The flume dual label is the joint capability t+ and t- which doesn't allow you to control declassifying/endorsing separately.

I think a slightly better way to do this than what Flume does is to have a tag t_dec for declassifying t and t_end for endorsing t. Then upgrades are always allowed for either type of label.
You would then also need:
declassifying_channel_send(...
endorsing_channel_recv(...

As long as "upgrades" of data being moved are always allowed, I actually see no need to have a a send that endorses or a recv that declassifies. If we did need to declassify on recvs and endorse on sends, the API gets more difficult because we really would need a "downgrading" channel call which does both. BUT AFAICT it looks like we can cleanly separate declassification/endorsement privilege if you like this idea.

Reply:
Thanks for the detailed comments, everything you said made sense to me! Let me reply to some of your points individually.

## explicit downgrading send / receive

I thought about this, but since the Node can already inspect the label of the channel it is about to send things over, I think this can be emulated by an inspection plus a comparison to the current node label, and based on that deciding whether to do the send or not. Perhaps we can package it up as a single function in the Rust SDK, but I think it does not require to add a different version of send / receive to the ABI. But happy to revisit at some later point, let's keep discussing this offline.

## separate declassify / endorse capabilities

I actually started with this approach of them being separate (following the Flume model), but I realized that for most (all?) practical cases I would always grant both of them anyways, so it seems to just add a new field which would always be the same as another one. Do you have any use case in which it is necessary to keep them separate?

## separate declassifying and endorsing tags

Not sure I completely follow your point here, are you saying to split the principals so that we always talk about more specific ones in the context of either declassification or endorsement? If so, I guess this is similar to always using explicit integrity and confidentiality projections in Flam, rather then the underlying joint principal. Is this correct?

Reply:
### explicit downgrading send / receive
Your response makes sense. Loosely the choice is defaulting towards throwing an error if there is a chance of information release vs. more conveniently assuming the node wants to use all privileges it has.

I like the idea of implementing it one way in the ABI and emulating other variants in the SDK. I think separating the downgrading case from the non-downgrading case in the ABI might also make a formalism about the ABI easier, but we are maybe too far away from that direction to make this a priority. .

I won't keep pushing for one answer or the other, though.

### separate declassify / endorse capabilities
Given that confidentiality / integrity are separate labels in the first place, it seems awkward to not control downgrades for them separately as well. If you want to collapse the ability to downgrade to one object, why not also collapse labels to one object? (You can also map a product lattice onto a single lattice). 

I think the ability to separate conf/integ downgrading privilege is easy enough to add that you might as well just include it even without a concrete use-case, yet. I think folks will be imaginative enough to find a use for this. 

I think the second-best option would be to collapse labels to one object and have downgrading capability as the only other object. The midpoint where there are separate conf/integ labels, but only one downgrading object seems awkward to me. Maybe it's possible to translate from the (conf, integ, decl, endo) model to the (label, downgrade) model.

### separate declassifying and endorsing tags
I was trying to suggest we improve over the Flume model in the following way.

Flume has a capability called t+ and another called t-.
t+ is two things: the ability to declassify secrecy, and the ability to "upgrade" integrity. (Where upgrading integrity means going from trusted to untrusted. This is the direction that is always allowed).
t- is dual to this.

I think this design unnecessarily constrains upgrades which should just be allowed by default. To be more concrete, I think flows from [conf: {}, ... ] -> [conf: {alice}, ... ] should just always be allowed.

I think it is better to have:
t_decl: which just controls declassification (but not integrity upgrades which are now always allowed)
t_endo: controls endorsement (but not secrecy upgrades which are now always allowed)

Reply:
(focusing on **separate declassifying and endorsing tags** only)

> Flume has a capability called t+ and another called t-.
t+ is two things: the ability to declassify secrecy, and the ability to "upgrade" integrity. (Where upgrading integrity means going from trusted to untrusted. This is the direction that is always allowed).
t- is dual to this.

I thought t- is the ability to declassify secrecy, and t+ is the ability to endorse integrity? Or I clearly misunderstood something :)

BTW I think the current design in this PR already allows for movements up the lattice, because it uses the "flows to" relation between labels, so, say, [conf: {}, ... ] -> [conf: {alice}, ... ] is indeed allowed. Here is the implementation of the "flows to" relation between labels: 

https://github.com/project-oak/oak/blob/6312906a7f711b4ba4508a9dcd85bbadd8c78b6f/oak/server/rust/oak_abi/src/label/mod.rs#L57-L71